src/linear_algebra/lagrange.lean

```diff
@@ -1,224 +1,567 @@
+  (basis s v i).nat_degree = s.card - 1 :=
 begin
-  convert degree_interpolate_lt (s.erase x) f,
-  rw finset.card_erase_of_mem hx,
+  have H : ∀ j, j ∈ s.erase i → basis_divisor (v i) (v j) ≠ 0,
+  { simp_rw [ne.def, mem_erase, basis_divisor_eq_zero_iff],
+    exact λ j ⟨hij₁, hj⟩ hij₂, hij₁ (hvs hj hi hij₂.symm) },
+  rw [← card_erase_of_mem hi, card_eq_sum_ones],
+  convert nat_degree_prod _ _ H using 1,
+  refine sum_congr rfl (λ j hj, (nat_degree_basis_divisor_ne _).symm),
+  rw [ne.def, ← basis_divisor_eq_zero_iff],
+  exact H _ hj
 end
 
-theorem interpolate_eq_of_eval_eq (f g : F → F) {s : finset F} (hs : ∀ x ∈ s, f x = g x) :
-  interpolate s f = interpolate s g :=
+theorem degree_basis (hvs : set.inj_on v s) (hi : i ∈ s) :
+  (basis s v i).degree = ↑(s.card - 1) :=
+by rw [degree_eq_nat_degree (basis_ne_zero hvs hi), nat_degree_basis hvs hi]
+
+lemma sum_basis (hvs : set.inj_on v s) (hs : s.nonempty) : ∑ j in s, (basis s v j) = 1 :=
 begin
-  rw [interpolate, interpolate],
-  refine finset.sum_congr rfl (λ x hx, _),
-  rw hs x hx,
+  refine eq_of_degrees_lt_of_eval_index_eq s hvs (lt_of_le_of_lt (degree_sum_le _ _) _) _ _,
+  { rw finset.sup_lt_iff (with_bot.bot_lt_coe s.card),
+    intros i hi,
+    rw [degree_basis hvs hi, with_bot.coe_lt_coe],
+    exact nat.pred_lt (card_ne_zero_of_mem hi) },
+  { rw [degree_one, ← with_bot.coe_zero, with_bot.coe_lt_coe],
+    exact nonempty.card_pos hs },
+  { intros i hi,
+    rw [eval_finset_sum, eval_one, ← add_sum_erase _ _ hi, eval_basis_eq hvs hi, add_right_eq_self],
+    refine sum_eq_zero (λ j hj, _),
+    rcases mem_erase.mp hj with ⟨hij, hj⟩,
+    rw eval_basis_ne hij hi }
 end
 
-/-- Linear version of `interpolate`. -/
-def linterpolate : (F → F) →ₗ[F] polynomial F :=
-{ to_fun := interpolate s,
-  map_add' := λ f g, by { simp_rw [interpolate, ← finset.sum_add_distrib, ← add_mul, ← C_add],
-    refl },
-  map_smul' := λ c f, by { simp_rw [interpolate, finset.smul_sum, C_mul', smul_smul], refl } }
+lemma basis_divisor_add_symm {x y : F} (hxy : x ≠ y) : basis_divisor x y + basis_divisor y x = 1 :=
+begin
+  classical,
+  rw [←sum_basis (set.inj_on_of_injective function.injective_id _) ⟨x, mem_insert_self _ {y}⟩,
+      sum_insert (not_mem_singleton.mpr hxy), sum_singleton, basis_doubleton_left hxy,
+      basis_doubleton_right hxy, id, id]
+end
+
+end basis
+
+section interpolate
+open finset
+variables {ι : Type*} [decidable_eq ι] {s t : finset ι} {i j : ι} {v : ι → F} (r r' : ι → F)
 
-@[simp] lemma interpolate_add (f g) : interpolate s (f + g) = interpolate s f + interpolate s g :=
-(linterpolate s).map_add f g
+/-- Lagrange interpolation: given a finset `s : finset ι`, a nodal map  `v : ι → F`
+which is injective on s, and a value function `r : ι → F`, `interpolate x r` is the unique
+polynomial of degree `< s.card` that takes value `r i` on `v i` for all `i` in `s`. -/
+def interpolate (s : finset ι) (v : ι → F) : (ι → F) →ₗ[F] F[X] :=
```

Comment:
The `simps` attribute should be able to automatically generate `interpolate_apply`:
```suggestion
@[simps]
def interpolate (s : finset ι) (v : ι → F) : (ι → F) →ₗ[F] F[X] :=
```